src/Management/test/Endpoint.Test/Health/HealthEndpointCoreTest.cs

```diff
@@ -28,11 +29,20 @@ public void Constructor_ThrowsOnNulls()
     {
         var contributors = new List<IHealthContributor>();
 
-        var optionsEx = Assert.Throws<ArgumentNullException>(() => new HealthEndpointCore(null, _aggregator, contributors, ServiceOptions(), _provider));
-        var aggregatorEx = Assert.Throws<ArgumentNullException>(() => new HealthEndpointCore(_options, null, contributors, ServiceOptions(), _provider));
-        var contribEx = Assert.Throws<ArgumentNullException>(() => new HealthEndpointCore(_options, _aggregator, null, ServiceOptions(), _provider));
-        var svcOptsEx = Assert.Throws<ArgumentNullException>(() => new HealthEndpointCore(_options, _aggregator, contributors, null, _provider));
-        var providerEx = Assert.Throws<ArgumentNullException>(() => new HealthEndpointCore(_options, _aggregator, contributors, ServiceOptions(), null));
+        var optionsEx = Assert.Throws<ArgumentNullException>(() =>
+            new HealthEndpointCore(null, _aggregator, contributors, ServiceOptions(), _provider, NullLogger<HealthEndpointCore>.Instance));
+
+        var aggregatorEx = Assert.Throws<ArgumentNullException>(() =>
+            new HealthEndpointCore(_options, null, contributors, ServiceOptions(), _provider, NullLogger<HealthEndpointCore>.Instance));
```

Comment:
I wonder if the `ILogger<HealthEndpointCore> logger` parameter on `HealthEndpointCore` should be `ILogger` instead. Is this type non-sealed with protected and virtual members solely to make it possible to inherit from in our tests, or is the type intended to be inherited from by Steeltoe users?

Reply:
It is meant for inheritance by Steeltoe users and they will be forced to use ILogger<HealthEndpointCore> as a parameter to the base class, which imo should be ok. see TestHealthEndpoint in the tests ...

Reply:
Ok, let me try to make my point by inheriting from the base type:

```c#
public class CustomHealthEndpointCore : HealthEndpointCore
{
    private readonly ILogger<HealthEndpointCore> _logger;

    public CustomHealthEndpointCore(/* other ctor parameters left out for brevity */
        ILogger<HealthEndpointCore> logger)
        : base(options, aggregator, contributors, serviceOptions, provider, logger)
    {
        _logger = logger;
    }

    public override HealthEndpointResponse Invoke(ISecurityContext securityContext)
    {
        _logger.LogDebug("Invoke derived.");
        return base.Invoke(securityContext);
    }
}
```

And for demonstration purposes, let's insert a log line to the base method:
```c#
    public virtual HealthEndpointResponse Invoke(ISecurityContext securityContext)
    {
        _logger.LogDebug("Invoke base.");
        return BuildHealth(securityContext);
    }
```

This would log:
```
HealthEndpointCore: Invoke derived.
HealthEndpointCore: Invoke base.
```

This doesn't look right, because the first log statement occurred in `CustomHealthEndpointCore` instead of `HealthEndpointCore`. To fix that, we'll need to add an extra constructor parameter with our own typed `ILogger`:

```c#
public class CustomHealthEndpointCore : HealthEndpointCore
{
    private readonly ILogger<CustomHealthEndpointCore> _customLogger;

    public CustomHealthEndpointCore(/* other ctor parameters left out for brevity */
        ILogger<HealthEndpointCore> logger,
        ILogger<CustomHealthEndpointCore> customLogger)
        : base(options, aggregator, contributors, serviceOptions, provider, logger)
    {
        _customLogger = customLogger;
    }

    public override HealthEndpointResponse Invoke(ISecurityContext securityContext)
    {
        _customLogger.LogDebug("Invoke derived.");
        return base.Invoke(securityContext);
    }
}
```

This would log:
```
CustomHealthEndpointCore: Invoke derived.
HealthEndpointCore: Invoke base.
```

This is correct, but it's a bit annoying that every derived type needs to add its own typed `ILogger` parameter. I suspect that's why many other Steeltoe types take an untyped `ILogger` parameter. For example, `AvailabilityHealthContributor` and `DiagnosticObserver`.

The downside of that approach is that logging statements from the base type now become logged under the name of the most-derived type. To let every type in the hierarchy log under its own name without adding many constructor parameters, each type in the hierarchy should take an `ILoggerFactory` instead. Then each type can create its own typed logger, so that it becomes easy to trace back which class a line originated from.

However, using an untyped `ILogger` in abstract types seems to be an established pattern in the Steeltoe codebase (I'm seeing hundreds of usages), so it seems to me we should stick to that.